aten/src/ATen/native/LinearAlgebra.cpp

```diff
@@ -2334,7 +2383,205 @@ Tensor matrix_exp_backward(const Tensor& self, const Tensor& grad) {
   );
 }
 
+TORCH_IMPL_FUNC(linalg_vector_norm_out)(const Tensor& self, const Scalar& scalar_ord, OptionalIntArrayRef opt_dim, bool keepdim, optional<ScalarType> opt_dtype, const Tensor& result) {
+  // Casting a large integer to a double will just introduce an error for
+  // values larger than 10^53 (same for negative numbers), so that's fine.
+  auto ord = scalar_ord.toDouble();
+  auto dim = opt_dim.value_or(IntArrayRef{});
+  // No need to handle opt_dtype explicitly as it is already encoded in the dtype of result
+
+  // Issue arising from the difference between vectorized and non-vectorized implementation on CPU
+  Tensor self_;
+  if (self.device().type() == c10::kCPU &&
+      isComplexType(self.scalar_type()) &&
+      std::abs(ord) == INFINITY) {
+    // TODO: This at::abs() call is used so that the at::abs() call in the
+    // backward function produces an identical result for complex inputs.
+    // However, it would be ideal if we could incorporate this into
+    // linalg_vector_norm_stub. See issue:
+    // https://github.com/pytorch/pytorch/issues/52648
+    auto in_dtype = opt_dtype.value_or(self.scalar_type());
+    self_ = self.to(in_dtype).abs();
+  } else {
+    self_ = self;
+  }
+
+  auto iter = make_reduction("vector_norm", const_cast<Tensor&>(result), self_, dim, keepdim, result.scalar_type());
+  norm_stub(iter.device_type(), iter, ord);
+}
+
+void _linalg_matrix_norm_checks(const Tensor& A, IntArrayRef dim, optional<ScalarType> opt_dtype) {
+  at::native::checkFloatingOrComplex(A, "linalg.matrix_norm");
+  TORCH_CHECK(A.dim() >= 2,
+              "linalg.matrix_norm: input tensor must be a matrix or a batch of matrices");
+
+  // dim
+  TORCH_CHECK(dim.size() == 2, "linalg.matrix_norm: dim must be a 2-tuple of ints");
+  TORCH_CHECK(dim[0] != dim[1], "Expected dims to be different, got (", dim[0], ", ", dim[1], ") instead");
+
+  // dtype
+  at::detail::check_linalg_norm_dtype(opt_dtype, A.scalar_type(), "linalg.matrix_norm");
+}
+
+Tensor linalg_matrix_norm(
+    const Tensor& A,
+    const Scalar& scalar_ord,
+    IntArrayRef dim,
+    bool keepdim,
+    optional<ScalarType> opt_dtype) {
+  _linalg_matrix_norm_checks(A, dim, opt_dtype);
+
+  auto ord = scalar_ord.toDouble();
+  auto abs_ord = std::abs(ord);
+  TORCH_CHECK(abs_ord == 2. || abs_ord == 1. || abs_ord == INFINITY, "linalg.matrix_norm: Order ", ord, " not supported.");
+
+  auto dim_ = dim.vec();
+  maybe_wrap_dims(dim_, A.dim());
+
+  auto max_min = [ord, keepdim](const Tensor& A, int64_t dim) { return ord > 0 ? A.amax(dim, keepdim) : A.amin(dim, keepdim); };
+  if (abs_ord == 2.) {
```

Comment:
For prims should we model abs_ord == 2 vs other ords as distinct prims that matrix_norm would call into?

Reply:
Natalia suggests matrix_norm should be a prim, anyway, so the way I asked this question is moot

Reply:
I agree with Natalia on this one, given that this operation is already modelled after NumPy (i.e. it's somewhat "standard").